src/_core/Settings.js

```diff
@@ -18,104 +15,130 @@ class Settings {
+          items[section][key]["description"] = description;
+          items[section][key]["id"] = id;
+          chrome.storage.sync.set(items, () => {
+            HFX.Logger.debug(`Added ${key} AND ${section}`);
+            HFX.Settings.proceedQueue(section);
+            cb();
+          });
+        } else {
+          items[section][key] = {};
+          items[section][key]["default"] = defaultOpt;
+          items[section][key]["enabled"] = defaultOpt;
+          items[section][key]["name"] = name;
+          items[section][key]["description"] = description;
+          items[section][key]["id"] = id;
+          chrome.storage.sync.set(items, () => {
+            HFX.Logger.debug(`Added ${key} in ${section}`);
+            HFX.Settings.proceedQueue(section);
+            cb();
+          });
+        }
+      } else if (purpose === "update") {
+        const key = queue[section].items[0].key;
+        const setting = queue[section].items[0].setting;
+        const value = queue[section].items[0].value;
+
+        items[section][key][setting] = value;
+        chrome.storage.sync.set(items, () => {
+          HFX.Logger.debug(`Updated ${key}:${setting}`);
+          HFX.Settings.proceedQueue(section);
         });
       }
     });
   }
 
-  proceedQueue() {
-    this.running = false;
-    queue.shift();
-    this.processQueue();
+  proceedQueue (section) {
+    queue[section].running = false;
+    queue[section].items.shift();
+    this.processQueue(section);
   }
 
-  printSettings() {
-    chrome.storage.sync.get(null, function (items) {
+  printSettings () {
+    chrome.storage.sync.get(null, (items) => {
       HFX.Logger.debug("Items: ", items);
     });
   }
 
-  exists(section, key, setting, cb) {
-    chrome.storage.sync.get(section, function (items) {
-      if (setting === null) {
-        return cb(Boolean(items[section][key]));
-      } else {
-        return cb(Boolean(items[section][key][setting]));
-      }
+  exists (section, key, setting, cb) {
+    chrome.storage.sync.get(section, (items) => {
+      return cb(HFX.Util.hasOwnPropertyStructure(items, section, key, setting));
     });
   }
 
-  get(section, key, setting, cb) {
-    chrome.storage.sync.get(section, function (items) {
-      if (typeof items[section][key] === 'undefined' || typeof items[section][key][setting] === 'undefined') {
-        return cb(null);
-      }
-      return cb(items[section][key][setting]);
+  get (section, key, setting, cb) {
+    chrome.storage.sync.get(section, (items) => {
+      return cb(HFX.Util.hasOwnPropertyStructure(items, section, key, setting) ? items[section][key][setting] : null);
     });
   }
 
-  set(section, key, setting, value) {
-    chrome.storage.sync.get(section, function (items) {
-      var obj = items;
-      obj[section][key][setting] = value;
-      chrome.storage.sync.set(obj, function () {
+  set (section, key, setting, value) {
+    HFX.Logger.warn("Unsafe storage updating.");
```

Comment:
What makes this unsafe storage?

Reply:
If something else makes a concurrent `set` or is in the `processQueue`, the value set by this call can be overwritten. This probably is unlikely to occur if everything is moved to their own sections (not under global), but it still would be possible if you do two `set` calls close enough together.

In particular, this was happening to the Alert module. Occasionally it would fetch the new alert, then forget the alert because something overwrote the setting.